connection_pool.go

```diff
@@ -98,9 +98,28 @@ func (b *backend) getNewConnection() (*connection, error) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
+	// Attempt to determine if we're over quota, if so, first start by cleaning out any
+	// connections that are closed, then recheck. We assert that the quota is only meant
+	// to count against open/in-use connections, so I don't care about closed ones.
 	if b.counter >= b.conf.ConnectionLimit {
-		go b.debugHitMaxConnections()
-		return nil, nil
+		newConns := make([]*connection, 0, b.counter)
+		for _, conn := range b.conns {
+			if !conn.IsClosed() {
+				newConns = append(newConns, conn)
+			}
+		}
+
+		// If the lengths are the same, we truly hit max connections and there's nothing
+		// to do so return
+		if len(newConns) == b.counter {
+			go b.debugHitMaxConnections()
+			return nil, nil
+		}
+
+		// We eliminated one or more closed connections, use the new list and reset our
```

Comment:
We could make the debug logic only remove connections the second time it's seen them, this will probably eliminate the race condition but means we get rid of connections less quickly and we have to add a counter.

I'm happy to do that if you think it's worth it?